sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/AlterTableDropPartitionSuite.scala

```diff
@@ -42,6 +42,18 @@ trait AlterTableDropPartitionSuiteBase extends command.AlterTableDropPartitionSu
       checkPartitions(t) // no partitions
     }
   }
+
+  test("SPARK-33941: invalidate cache after partition dropping") {
+    withNamespaceAndTable("ns", "tbl") { t =>
+      sql(s"CREATE TABLE $t (id int, part int) $defaultUsing PARTITIONED BY (part)")
+      sql(s"INSERT INTO $t PARTITION (part=0) SELECT 0")
+      val df = spark.table(t)
+      df.cache()
+      assert(!df.isEmpty)
+      sql(s"ALTER TABLE $t DROP PARTITION (part=0)")
+      assert(df.isEmpty)
```

Comment:
can we check that the cache still exist after the command? this will validate that the cache is actually refreshed rather than removed.

Reply:
I still use `df.isEmpty` instead of checking the content of the dataframe because `sparkSession.catalog.refreshTable` is not enough, it seems. It doesn't refresh the file indexes for the table.

So, test fails:
```
  test("SPARK-33941: refresh cache after partition dropping") {
    withNamespaceAndTable("ns", "tbl") { t =>
      sql(s"CREATE TABLE $t (id int, part int) $defaultUsing PARTITIONED BY (part)")
      sql(s"INSERT INTO $t PARTITION (part=0) SELECT 0")
      sql(s"INSERT INTO $t PARTITION (part=1) SELECT 1")
      val df = spark.table(t)
      df.cache()
      df.collect()
      sql(s"ALTER TABLE $t DROP PARTITION (part=0)")
      df.collect()
    }
  }
```
with the exception:
```
java.io.FileNotFoundException: File file:/Users/maximgekk/proj/drop-partition-invalidate-cache/spark-warehouse/org.apache.spark.sql.execution.command.v1.AlterTableDropPartitionSuite/ns.db/tbl/part=0/part-00000-4536ff40-a1ac-41f4-9ff5-7fbc7d67c7b9.c000.snappy.parquet does not exist
It is possible the underlying files have been updated. You can explicitly invalidate the cache in Spark by running 'REFRESH TABLE tableName' command in SQL or by recreating the Dataset/DataFrame involved.
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.org$apache$spark$sql$execution$datasources$FileScanRDD$$anon$$readCurrentFile(FileScanRDD.scala:124)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:169)
```

And `REFRESH TABLE` after `ALTER TABLE .. DROP PARTITION` doesn't help since it doesn't refresh the file index too :-(

Reply:
hmm, even simpler test without caching fails:
```scala
  test("SPARK-33941: refresh cache after partition dropping") {
    withNamespaceAndTable("ns", "tbl") { t =>
      sql(s"CREATE TABLE $t (id int, part int) $defaultUsing PARTITIONED BY (part)")
      sql(s"INSERT INTO $t PARTITION (part=0) SELECT 0")
      sql(s"INSERT INTO $t PARTITION (part=1) SELECT 1")
      val df = spark.table(t)
      // df.cache()
      df.collect()
      sql(s"ALTER TABLE $t DROP PARTITION (part=0)")
      sql(s"ALTER TABLE $t RECOVER PARTITIONS")
      sql(s"MSCK REPAIR TABLE $t")
      df.collect()
    }
  }
```
with:
```java
Input path does not exist: file:/private/var/folders/p3/dfs6mf655d7fnjrsjvldh0tc0000gn/T/warehouse-a3f5be00-8e30-43be-a14d-a17983f2cd2d/ns.db/tbl/part=0
org.apache.hadoop.mapred.InvalidInputException: Input path does not exist: file:/private/var/folders/p3/dfs6mf655d7fnjrsjvldh0tc0000gn/T/warehouse-a3f5be00-8e30-43be-a14d-a17983f2cd2d/ns.db/tbl/part=0
	at org.apache.hadoop.mapred.FileInputFormat.singleThreadedListStatus(FileInputFormat.java:297)
	at org.apache.hadoop.mapred.FileInputFormat.listStatus(FileInputFormat.java:239)
```
on v1 In-Memory and Hive external catalogs. The issue looks much more serious than I could expect. cc @cloud-fan @HyukjinKwon

Reply:
Hmm interesting... I thought this should have been addressed via #30699 but maybe it's a different one. I'll check this when I have time too.